Implement `ReadableStream.from`

andreubotella opened the pull request:
#### 1eea8e2ede0e4b4428af42d91ef60fd10cb10ff1
<pre>
Implement `ReadableStream.from`
<a href="https://bugs.webkit.org/show_bug.cgi?id=255086">https://bugs.webkit.org/show_bug.cgi?id=255086</a>

Reviewed by NOBODY (OOPS!).

The `ReadableStream.from` static method allows creating a ReadableStream
from a JS iterable object, whether sync or async.

* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.js: Added.
(async asyncGenerator):
(const.it.next):
(return.new.ReadableStream.start):
(factory.of.iterableFactories.promise_test.async const):
(factory.of.iterableFactories.promise_test):
(iterable.of.badIterables.test):
(test):
(test.t.const.iterable.Symbol.asyncIterator):
(promise_test.async const.iterable.async next):
(promise_test.async const):
(promise_test):
(promise_test.async t.const.iterable.async next):
(promise_test.async t):
(promise_test.const.iterable.async next):
(promise_test.async let):
(promise_test.async t.const.iterable.async return):
(promise_test.const.iterable.async return):
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/w3c-import.log:
* Source/WebCore/Modules/streams/ReadableStream.cpp:
(WebCore::ReadableStream::from):
* Source/WebCore/Modules/streams/ReadableStream.h:
* Source/WebCore/Modules/streams/ReadableStream.idl:
* Source/WebCore/Modules/streams/ReadableStreamInternals.js:
(createInternalReadableStreamFromAsyncIterable):
* Source/WebCore/bindings/js/InternalReadableStream.cpp:
(WebCore::InternalReadableStream::createFromAsyncIterable):
* Source/WebCore/bindings/js/InternalReadableStream.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1eea8e2ede0e4b4428af42d91ef60fd10cb10ff1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26331 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4942 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27591 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28429 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24100 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26666 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6730 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2362 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24098 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26589 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3792 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22643 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29007 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3383 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23615 "Found 2 new test failures: imported/w3c/web-platform-tests/streams/readable-streams/from.any.serviceworker.html, imported/w3c/web-platform-tests/streams/readable-streams/from.any.sharedworker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29676 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22954 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24005 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24013 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27569 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25545 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3414 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1623 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/windows/browsing-context-names/choose-_blank-001.html (failure)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/32990 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4837 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23362 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7133 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3891 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3737 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

andreubotella commented:
cc @youennf as well.

The two failing tests depend on `ReadableStream` being an async iterable, which is not yet implemented (https://bugs.webkit.org/show_bug.cgi?id=194379).

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6a48205f6d0f46b348d47f7272d96702e0cf42a7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12454

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/002f492348778c1421510e8ec16b22b4fffbda17. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12454

youennf commented on Source/WebCore/Modules/streams/ReadableStream.cpp:
```diff
@@ -65,6 +65,17 @@ Ref<ReadableStream> ReadableStream::create(Ref<InternalReadableStream>&& interna
 
+ExceptionOr<Ref<ReadableStream>> ReadableStream::from(JSDOMGlobalObject& globalObject, JSC::JSValue asyncIterable)
+{
+    auto& jsDOMGlobalObject = *JSC::jsCast<JSDOMGlobalObject*>(&globalObject);
```
globalObject is already a JSDOMGlobalObject.

youennf commented on Source/WebCore/Modules/streams/ReadableStream.cpp:
```diff
@@ -65,6 +65,17 @@ Ref<ReadableStream> ReadableStream::create(Ref<InternalReadableStream>&& interna
+ExceptionOr<Ref<ReadableStream>> ReadableStream::from(JSDOMGlobalObject& globalObject, JSC::JSValue asyncIterable)
+{
+    auto& jsDOMGlobalObject = *JSC::jsCast<JSDOMGlobalObject*>(&globalObject);
+    RefPtr protectedContext { jsDOMGlobalObject.scriptExecutionContext() };
```
I am not clear why we are protecting context? Can you clarify?

youennf commented on Source/WebCore/Modules/streams/ReadableStream.idl:
```diff
@@ -35,6 +35,8 @@
 ] interface ReadableStream {
     [CallWith=CurrentGlobalObject] constructor(optional object underlyingSource, optional object options);
 
+    [CallWith=CurrentGlobalObject] static ReadableStream from(any asyncIterable);
```
We should probably add a runtime flag off by default, except for tests.

youennf reviewed (commented):
I need to look more closely at createInternalReadableStreamFromAsyncIterable.
Some feedback already.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5fba627e752eee350fc766fec0db1bcd6fa56eb8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12454

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8cfcca62de4512fe9a38012de75beaad62d2cfdb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12454

andreubotella commented:
I don't know why the mac-wk{1,2} tests keep failing. I made sure to run the relevant failing tests in a Mac machine, rather than just testing locally with the GTK port, and they run successfully.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/95ef5cfa4c72b7cc0a9912f6e42c4e1e3cd85a22. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12454

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c2bcc71889d3c19b0c5b14aaa86ca1f6ebb99500. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12454

andreubotella committed:
Implement `ReadableStream.from`
https://bugs.webkit.org/show_bug.cgi?id=255086

Reviewed by NOBODY (OOPS!).

The `ReadableStream.from` static method allows creating a ReadableStream
from a JS iterable object, whether sync or async.

* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.js: Added.
(async asyncGenerator):
(const.it.next):
(return.new.ReadableStream.start):
(factory.of.iterableFactories.promise_test.async const):
(factory.of.iterableFactories.promise_test):
(iterable.of.badIterables.test):
(test):
(test.t.const.iterable.Symbol.asyncIterator):
(promise_test.async const.iterable.async next):
(promise_test.async const):
(promise_test):
(promise_test.async t.const.iterable.async next):
(promise_test.async t):
(promise_test.const.iterable.async next):
(promise_test.async let):
(promise_test.async t.const.iterable.async return):
(promise_test.const.iterable.async return):
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.serviceworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.sharedworker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/from.any.worker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/streams/readable-streams/w3c-import.log:
* Source/WebCore/Modules/streams/ReadableStream.cpp:
(WebCore::ReadableStream::from):
* Source/WebCore/Modules/streams/ReadableStream.h:
* Source/WebCore/Modules/streams/ReadableStream.idl:
* Source/WebCore/Modules/streams/ReadableStreamInternals.js:
(createInternalReadableStreamFromAsyncIterable):
* Source/WebCore/bindings/js/InternalReadableStream.cpp:
(WebCore::InternalReadableStream::createFromAsyncIterable):
* Source/WebCore/bindings/js/InternalReadableStream.h:

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1eea8e2ede0e4b4428af42d91ef60fd10cb10ff1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12454
